sklearn/base.py

```diff
@@ -92,10 +92,20 @@ def clone(estimator, *, safe=True):
     for name in new_object_params:
         param1 = new_object_params[name]
         param2 = params_set[name]
-        if param1 is not param2:
+        init_param_error = (
+            f"Cannot clone object {estimator}, as the constructor "
+            f"either does not set or modifies parameter {name}."
+        )
+        if param1 is param2:
+            continue
+        try:
+            if param1 != param2:
+                raise RuntimeError(init_param_error)
```

Comment:
Can we inline the error message here so we do not construct the string when there is no error?